compiler/rustc_mir/src/borrow_check/mod.rs

```diff
@@ -105,7 +105,7 @@ fn mir_borrowck<'tcx>(
     let (input_body, promoted) = tcx.mir_promoted(def);
     debug!("run query mir_borrowck: {}", tcx.def_path_str(def.did.to_def_id()));
 
-    let opt_closure_req = tcx.infer_ctxt().enter(|infcx| {
+    let opt_closure_req = tcx.infer_ctxt().with_opaque_type_inference(def.did).enter(|infcx| {
```

Comment:
Not sure I love this approach but unsure neither how to make it better. One could forget calling `with_opaqie_type_inference`, would rather have a way to force us pass the `def_id` down always.

Reply:
Forgetting is safe, the thing just stops being a defining use. We can always add more sites later.

We should also only ever do this in mir borrowck's "type check" pass and in the real type check pass, so both are covered. I will document this information in the function's docs.